feature-dapp-impl/src/main/java/io/novafoundation/nova/feature_dapp_impl/domain/DappInteractor.kt

```diff
@@ -51,25 +54,44 @@ class DappInteractor(
             .map { it.sortDApps() }
     }
 
-    fun observeDAppsByCategory(): Flow<GroupedList<DappCategory, DApp>> {
+    fun observeDAppsByCategory(): Flow<DAppGroupedCatalog> {
+        val shufflingSeed = Random.nextInt()
+
         return combine(
             dAppMetadataRepository.observeDAppCatalog(),
             favouritesDAppRepository.observeFavourites()
         ) { dAppCatalog, favourites ->
+            // We use random with seed to shuffle dapps in categories the same way during updates
+            val random = Random(shufflingSeed)
+
             val categories = dAppCatalog.categories
             val dapps = dAppCatalog.dApps
 
             val urlToDAppMapping = buildUrlToDappMapping(dapps, favourites)
 
-            // Regrouping in O(Categories * Dapps)
-            // Complexity should be fine for expected amount of dApps
-            categories.associateWith { category ->
-                dapps.filter { category in it.categories }
-                    .map { urlToDAppMapping.getValue(it.url) }
-            }
+            val popular = dAppCatalog.popular.mapNotNull { urlToDAppMapping[it] }
+            val catalog = categories.associateWith { getShuffledDAppsInCategory(it, dapps, urlToDAppMapping, dAppCatalog.popular, random) }
+
+            DAppGroupedCatalog(popular, catalog)
         }
     }
 
+    private fun getShuffledDAppsInCategory(
+        category: DappCategory,
+        dapps: List<DappMetadata>,
+        urlToDAppMapping: Map<String, DApp>,
+        popular: List<DAppUrl>,
+        shufflingSeed: Random
+    ): List<DApp> {
+        val categoryDApps = dapps.filter { category in it.categories }
+            .map { urlToDAppMapping.getValue(it.url) }
+
+        val popularDAppsInCategory = categoryDApps.filter { it.url in popular }
+        val otherDAppsInCategory = categoryDApps.filterNot { it.url in popular }
```

Comment:
We can use
```
val (popularDAppsInCategory, otherDAppsInCategory) = categoryDApps.partition { it.url in popular } 
```

We also need to convert `popular` to `Set` before doing paritioning to avoid O(N^2) complexity